File: game/scripts/vscripts/modifiers/upgrades/modifier_upgrade_phantom_dash_shield.lua
modifier_upgrade_phantom_dash_shield = class({})
function modifier_upgrade_phantom_dash_shield:RemoveOnDeath() return false end
function modifier_upgrade_phantom_dash_shield:IsPurgable() return false end

function modifier_upgrade_phantom_dash_shield:OnCreated(params)
    if IsServer() then
        self.damage_block = 10
        self:SetStackCount(1)
    end
end

function modifier_upgrade_phantom_dash_shield:OnRefresh()
    if IsServer() then
        self:IncrementStackCount()
    end
end

function modifier_upgrade_phantom_dash_shield:GetDamageBlock(params)
    if IsServer() then
        return self:GetStackCount() * self.damage_block
    end
end

File: src/vscripts/modifiers/upgrades/shards/modifier_upgrade_phantom_dash_shield.ts
import { registerModifier } from "../../../lib/dota_ts_adapter";
import { ModifierUpgrade } from "../modifier_upgrade";

@registerModifier({ customNameForI18n: "modifier_upgrade_phantom_dash_shield" })
export class ModifierUpgradePhantomDashShield extends ModifierUpgrade {
    rangeBonus = 25;
    damageBlock = 10;

    OnCreated() {
        if (IsServer()) {
            this.SetStackCount(1);
        }
    }

    OnRefresh() {
        if (IsServer()) {
            this.IncrementStackCount();
        }
    }

    GetDamageBlock() {
        if (IsServer()) {
            return this.GetStackCount() * this.damageBlock;
        }

        return 0;
    }

    DeclareFunctions() {
        return [ModifierFunction.CAST_RANGE_BONUS_STACKING];
    }

    GetModifierCastRangeBonusStacking() {
        return this.GetStackCount() * this.rangeBonus;
    }
}
